### ISSUE TITLE
support for setting and verifying the Agent Policy hash value in TDX MRCONFIGID

### ISSUE THREAD

**Issue by danmihai1** [kata-containers/kata-containers#7669]
Which feature do you think can be improved?

Agent Policy (https://github.com/kata-containers/kata-containers/issues/7573)

How can it be improved?

- The Shim should be able to set the Policy hash value into the MRCONFIGID field of TDX Guest VMs.
- The Agent should be able to verify that the value of the MRCONFIGID field matches the Policy file it gets from the Shim.

Additional Information

- CoCo remote attestation should also check the expected value of the MRCONFIGID field.
- Additional details: https://github.com/confidential-containers/confidential-containers/issues/126#issuecomment-1644402591

Before raising this enhancement request

Have you looked at the [limitations document](https://github.com/kata-containers/kata-containers/blob/main/docs/Limitations.md)?
Yes

**Comment by Xynnn007** [comment]
Hi Dan, this concrete issue looks great to me. Based on the design I have some more thoughts to resolve other problems we're facing in CoCo.

Related to https://github.com/kata-containers/kata-containers/issues/7668 .

I think the way to leverage `MRCONFIGID` or `HOSTDATA` mechanism (let's call them "init data digest" then) should expose an easy and extensible API for users. Currently only policy is needed to be covered by HOSTDATA, but we can expect that the KBS public key cert and some other resources with the following attributes also need to be protected by init data:

1. The integrity needs to be protected while the confidentiality does not
2. should be measured but cannot influence the initial measurement like MRTD in TDX or `measurement` in SNP

Some init data examples
- KBS public key cert
- kernel commandline like aa_kbc_params which is specific to the current pod: Yes. For TDX it can be measured by MRTD. but for SNP, any change upon kernel commandline will change the final measurement if we do not use a vTPM inside guest firmware. 
- Launch parameters needed to be passed to CDH & AA when doing `CreateSandbox`

The easiest and most rationale way up to now is to extend the [SetPolicy](https://github.com/kata-containers/kata-containers/pull/7574/files#diff-b8ee3749d4b783b2fe64a459b83f01d26a23a6bb939310469afb2aad900545eaR75) API. Change the semantics to `SetInitData` and make it a single use API, which means that any kata-agent serve the call to api only once and then close.

Maybe `SetInitData` should be part of the atomic operation context of `CreateSandbox`. The steps would be like the following:
1. runtime receives annotation field which contains different init data from k8s yaml
2. runtime canonicalize the data, calculate init data digest and set as MRCONFIGID/HOSTDATA
3. launch CVM, call `SetInitData` to pass init data

Then do
> The Agent should be able to verify that the value of the MRCONFIGID field matches the Policy file it gets from the Shim.

or we can do the verification lazily, s.t. 
1. kata-agent calculates the digest of the init data it receives and record
2. when remote attestation occurs, extract HOSTDATA/MRCONFIGID field to compare with the one calculated in 1

cc @fitzthum @arronwy

**Comment by danmihai1** [comment]
Hi Ding,

Thanks for your thoughtful feedback - as always!

I didn't follow just this line:

>> 2. should be **measured** but cannot influence the initial measurement like MRTD in TDX or measurement in SNP

Did you mean: "should be **included in the hardware attestation report** but cannot influence the initial measurement...", or something else?

I agree that sending the KBS public key cert, aa_kbc_params, and similar data using the kernel command line is impractical.

All the init data examples you mentioned are in my mind part of the Guest "Secure Configuration" - a different name I've been using for the new Agent Policy. Therefore, I encourage you to think if including the KBS public key cert, aa_kbc_params, and similar data in a Rego policy file would be reasonable. Rego files can hold arbitrary data (e.g., in JSON format), so it should work for your init data examples too.

If using Rego text to send this data is unacceptable for some reason, then we could look at supporting multiple Policy formats.
For example: `SetPolicy(policy_type, policy_contents)` Rego would be one of the policy_types. But I really hope that Rego is acceptable for everyone in the foreseeable future, and we won't add more complexity with other policy types.

Otherwise, I wouldn't mind changing the API name from SetPolicy to SetInitData or similar. But again, in my mind "Guest init data" sounds like a section of a Guest Policy file, so SetPolicy seems like a reasonable API name.

I will think more about this, but I wanted to give you my initial reaction.

Thanks!

@Xynnn007 @fitzthum @arronwy

**Comment by Xynnn007** [comment]
> Did you mean: "should be included in the hardware attestation report but cannot influence the initial measurement...", or something else?

Yes. should be included in the hardware attestation report (or digest should be binded to the report)


about whether we should use rego to carry such configs, my first idea is no as this would make some coupling to the inputs. I also need to think about this carefully.

**Comment by arronwy** [comment]
Yes, fully agree we need ensure the integrity of policy file for CoCo. Use `MRCONFIGID` or `HOSTDATA` is a good option. It also sounds good for me that @danmihai1 's suggestion to combine policy file and initial data together, we can package them together as an OPA bundle tar.gz file and initial data can in the `data.json` file. The final gap is how to easily for the user to write or auto generate this OPA bundle file and send to guest VM, one potential option is save the bundle file in K8S configmap?(but we can not bind configmap to pod, need some workaround in kata runtime :-( ) , during pod launch, write the hash of opa bundle to `MRCONFIGID` or `HOSTDATA`  and kata runtime get the bundle file from k8s configmap? pass this file to guest with kata runtime @fidencio  @mkbhanda any suggestions?

**Comment by fitzthum** [comment]
I think we're all on the same page for implementing the measurement of the policy, but I think things are a bit more complicated when we start thinking about using the policy file for additional configurations or trying to use host data / MRCONFIG for additional configuration. I guess that might be a bit beyond the scope of the original issue, but this seems like a good place to discuss anyway. My thoughts are a bit disorganized here, so let me just mention some things. 

For policy, I think we have a nice design where platforms that have hostdata can use that to pass in the policy file, but the policy can also be retrieved from a KBS and it can also come from the FS directly. Once we start talking about adding other configurations, this gets more complicated. For example, if we want to put configurations relating to the AA/CDH into the policy config, then we won't really be able to use those components to fetch the policy config. I don't think that is quite what is being proposed, but it's an example of the sort of thing we should watch out for.

I guess different platform capabilities are one of the big issues here. I am wary of banking too much on configurations that are measured via hostdata, because that isn't universal. That said, it might be reasonable for certain platforms to not support every feature. Either way I think we should be really deliberate about what is supported and what is not. Probably a reasonable approach for a config is to have something in the initrd that we can fall back to, but this won't really work for configurations that are highly dynamic.

I also think we need to be vigilant about the experience of the client. Currently you can deploy CoCo just by adding a runtime class to your yaml file. Obviously more advanced features will require more configuration, but we should try to be thoughtful about presenting the client with something that is easy to use.

I realize that is a bit vague. I am looking forward to seeing more concretely how we try to integrate all these new pieces together. It might take us a couple iterations to get something that is really nice.

**Comment by danmihai1** [comment]
Perhaps we need to discuss these use cases in future meetings. Meanwhile, I am summarizing my current thoughts:

1. **For TEEs that support HOST_DATA/MRCONFIGID or similar field**, the Guest init data such as the KBS public key cert can be included in the pod Policy and sent to the Guest using SetPolicy().

2. **For TEEs that don't support HOST_DATA/MRCONFIGID or similar field**, my current understanding is that customer's root of trust for cloud services (is that the KBS public key cert?) must be included in the integrity protected Guest initrd or img, or passed to the Guest in some other way is that is measured by the TEE.

For (2), Guest init data such as the KBS public key cert can be included in the default policy file.

For both (1) and (2), the policy is available early during Guest boot, and can be retrieved by Guest components from OPA.

**Comment by fitzthum** [comment]
I think the above is generally fine, but I don't think it accounts for a certain type of configuration, which is a mandatory measured configuration that is dynamic. For example, there has been some discussion of configuring the resources that are requested by the CDH via this kind of config. That kind of config is somewhat likely to change depending on the workload (although I guess a client could request certain things like their KMS credentials across many different workloads). I think (2) is fine for things like the API settings for the proxy (which is scoped to the runtime class), but for workload dependent configs I'm not so sure.

Of course there aren't many other options. Putting things in the kernel command line doesn't really simplify anything. So it might be inevitable that these values are hard to measure on certain platforms. That could be a bit of a problem if the CDH relies on them for basic functionality. 

So some concrete questions: how dynamic are the configs for the CDH? Could it work for clients to just set them once in their initrd? It still is not optimal for a client to need to update the initrd at all, but maybe it could work. Are we sure that all the CDH configs actually need to be measured? Is there actually any security risk to someone changing the name of the secrets we're requesting? Won't that just break the workload?

**Comment by Xynnn007** [comment]
AFAIC what configs are for CDH is not important. Currently we have two layer attestation phase. Stage 1 attestation service will check the generic TCB of the software and hardware evidence and returns a signed parsed claims. Stage two is that the OPA in an authorization service (like current KBS that surely authorizes the guest to access the resources to the KBS). In this context, the things in init data like CDH config will be checked by stage 2 authorization service. I think the identity mark can also be included in init data and checked by authorization service. One thing to mention: the core meaning of remote attestation is authorize the guest to access confidential resources. If no confidential resource accesses are needed, RA is meaningless.

Not related to this issue, we should decouple the KBS into different modules due to different functionalities: authorization module that provisions token/session, confidential resource storage.

> Is there actually any security risk to someone changing the name of the secrets we're requesting? Won't that just break the workload?

the secret name/signer should also filtered by OPA policy, which requires us to define new rules in the in-guest policy.

**Comment by burgerdev** [comment]
1. We deprecated setting policy through `agent.policy` annotations / `SetPolicy` method, in favour of initdata.
2. The initdata hash is set to HOSTDATA / MRCONFIGID [by the runtime](https://github.com/kata-containers/kata-containers/blob/74254cba8f299e0bd76a60fd7e0da7cbeaf4b29f/src/runtime/pkg/govmm/qemu/qemu.go#L408-L439).
3. The hash is then verified by guest components ([current](https://github.com/kata-containers/kata-containers/blob/main/src/agent/src/main.rs#L417) / [future](https://github.com/kata-containers/kata-containers/issues/11532)).

So, I believe the work here is done.
